File: fortran/matrix_operations.f90
module matrix_operations
    implicit none

contains

    subroutine matrix_multiply(A, B, C, n, m, p)
        implicit none
        integer, intent(in) :: n, m, p
        real(kind=8), intent(in) :: A(n, m), B(m, p)
        real(kind=8), intent(out) :: C(n, p)
        integer :: i, j, k

        C = 0.0d0
        do i = 1, n
            do j = 1, p
                do k = 1, m
                    C(i, j) = C(i, j) + A(i, k) * B(k, j)
                end do
            end do
        end do
    end subroutine matrix_multiply

    subroutine lu_decomposition(A, L, U, n)
        implicit none
        integer, intent(in) :: n
        real(kind=8), intent(in) :: A(n, n)
        real(kind=8), intent(out) :: L(n, n), U(n, n)
        integer :: i, j, k
        real(kind=8) :: sum_val

        L = 0.0d0
        U = 0.0d0

        do i = 1, n
            do j = i, n
                sum_val = 0.0d0
                do k = 1, i - 1
                    sum_val = sum_val + L(i, k) * U(k, j)
                end do
                U(i, j) = A(i, j) - sum_val
            end do

            do j = i + 1, n
                sum_val = 0.0d0
                do k = 1, i - 1
                    sum_val = sum_val + L(j, k) * U(k, i)
                end do
                if (dabs(U(i, i)) < 1.0d-15) then
                    L(j, i) = 0.0d0
                else
                    L(j, i) = (A(j, i) - sum_val) / U(i, i)
                end if
            end do
        end do

        do i = 1, n
            L(i, i) = 1.0d0
        end do
    end subroutine lu_decomposition

    subroutine gaussian_elimination(A, b, x, n)
        implicit none
        integer, intent(in) :: n
        real(kind=8), intent(inout) :: A(n, n), b(n)
        real(kind=8), intent(out) :: x(n)
        integer :: i, j, k
        real(kind=8) :: factor, temp
        integer :: pivot_row

        ! Forward elimination
        do k = 1, n - 1
            ! Partial pivoting
            pivot_row = k
            do i = k + 1, n
                if (dabs(A(i, k)) > dabs(A(pivot_row, k))) then
                    pivot_row = i
                end if
            end do

            ! Swap rows if needed
            if (pivot_row /= k) then
                do j = k, n
                    temp = A(k, j)
                    A(k, j) = A(pivot_row, j)
                    A(pivot_row, j) = temp
                end do
                temp = b(k)
                b(k) = b(pivot_row)
                b(pivot_row) = temp
            end if

            ! Eliminate below pivot
            do i = k + 1, n
                if (dabs(A(k, k)) > 1.0d-15) then
                    factor = A(i, k) / A(k, k)
                    do j = k, n
                        A(i, j) = A(i, j) - factor * A(k, j)
                    end do
                    b(i) = b(i) - factor * b(k)
                end if
            end do
        end do

        ! Back substitution
        do i = n, 1, -1
            x(i) = b(i)
            do j = i + 1, n
                x(i) = x(i) - A(i, j) * x(j)
            end do
            if (dabs(A(i, i)) > 1.0d-15) then
                x(i) = x(i) / A(i, i)
            end if
        end do
    end subroutine gaussian_elimination

    subroutine print_matrix(A, n, m, name)
        implicit none
        integer, intent(in) :: n, m
        real(kind=8), intent(in) :: A(n, m)
        character(len=*), intent(in) :: name
        integer :: i, j

        write(*, '(A)') trim(name)
        do i = 1, n
            do j = 1, m
                write(*, '(F12.6)', advance='no') A(i, j)
            end do
            write(*, '(A)') ''
        end do
        write(*, '(A)') ''
    end subroutine print_matrix

end module matrix_operations

program matrix_ops_test
    use matrix_operations
    implicit none

    integer, parameter :: n = 3
    real(kind=8) :: A(n, n), B(n, n), C(n, n)
    real(kind=8) :: L(n, n), U(n, n)
    integer :: i, j

    ! Initialize test matrices
    A = reshape([4.0d0, 3.0d0, 1.0d0, &
                 3.0d0, 5.0d0, 2.0d0, &
                 1.0d0, 2.0d0, 3.0d0], shape=[n, n])

    B = reshape([1.0d0, 0.0d0, 0.0d0, &
                 0.0d0, 1.0d0, 0.0d0, &
                 0.0d0, 0.0d0, 1.0d0], shape=[n, n])

    ! Test matrix multiplication
    call matrix_multiply(A, B, C, n, n, n)
    call print_matrix(C, n, n, "Matrix A * B:")

    ! Test LU decomposition
    call lu_decomposition(A, L, U, n)
    call print_matrix(L, n, n, "L matrix from LU decomposition:")
    call print_matrix(U, n, n, "U matrix from LU decomposition:")

    write(*, '(A)') "====================================="
    write(*, '(A)') "Matrix Operations Complete"
    write(*, '(A)') "====================================="
end program matrix_ops_test

File: src/matrix_operations.rs
/// Matrix Operations Module
/// Rust translation of Fortran matrix_operations.f90
/// Implements matrix multiplication, LU decomposition, and Gaussian elimination

pub struct MatrixOps;

impl MatrixOps {
    /// Matrix multiplication: C = A * B
    pub fn matrix_multiply(a: &[Vec<f64>], b: &[Vec<f64>]) -> Vec<Vec<f64>> {
        let n = a.len();
        let m = a[0].len();
        let p = b[0].len();

        let mut c = vec![vec![0.0; p]; n];

        for i in 0..n {
            for j in 0..p {
                for k in 0..m {
                    c[i][j] += a[i][k] * b[k][j];
                }
            }
        }

        c
    }

    /// LU Decomposition: A = L * U
    pub fn lu_decomposition(a: &[Vec<f64>]) -> (Vec<Vec<f64>>, Vec<Vec<f64>>) {
        let n = a.len();
        let mut l = vec![vec![0.0; n]; n];
        let mut u = vec![vec![0.0; n]; n];

        for i in 0..n {
            // Compute U row
            for j in i..n {
                let mut sum = 0.0;
                for k in 0..i {
                    sum += l[i][k] * u[k][j];
                }
                u[i][j] = a[i][j] - sum;
            }

            // Compute L column
            for j in i + 1..n {
                let mut sum = 0.0;
                for k in 0..i {
                    sum += l[j][k] * u[k][i];
                }

                if u[i][i].abs() > 1e-15 {
                    l[j][i] = (a[j][i] - sum) / u[i][i];
                } else {
                    l[j][i] = 0.0;
                }
            }
        }

        // Set diagonal of L to 1
        for i in 0..n {
            l[i][i] = 1.0;
        }

        (l, u)
    }

    /// Gaussian Elimination with partial pivoting
    pub fn gaussian_elimination(a_input: &[Vec<f64>], b_input: &[f64]) -> Vec<f64> {
        let n = a_input.len();
        let mut a = a_input.to_vec();
        let mut b = b_input.to_vec();

        // Forward elimination
        for k in 0..n - 1 {
            // Partial pivoting
            let mut pivot_row = k;
            for i in k + 1..n {
                if a[i][k].abs() > a[pivot_row][k].abs() {
                    pivot_row = i;
                }
            }

            // Swap rows if needed
            if pivot_row != k {
                for j in k..n {
                    let temp = a[k][j];
                    a[k][j] = a[pivot_row][j];
                    a[pivot_row][j] = temp;
                }
                let temp = b[k];
                b[k] = b[pivot_row];
                b[pivot_row] = temp;
            }

            // Eliminate below pivot
            for i in k + 1..n {
                if a[k][k].abs() > 1e-15 {
                    let factor = a[i][k] / a[k][k];
                    for j in k..n {
                        a[i][j] -= factor * a[k][j];
                    }
                    b[i] -= factor * b[k];
                }
            }
        }

        // Back substitution
        let mut x = vec![0.0; n];
        for i in (0..n).rev() {
            x[i] = b[i];
            for j in i + 1..n {
                x[i] -= a[i][j] * x[j];
            }
            if a[i][i].abs() > 1e-15 {
                x[i] /= a[i][i];
            }
        }

        x
    }

    pub fn print_matrix(a: &[Vec<f64>], name: &str) {
        println!("{}", name);
        for row in a {
            for val in row {
                print!("{:12.6} ", val);
            }
            println!();
        }
        println!();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_matrix_multiply_identity() {
        let a = vec![vec![1.0, 0.0], vec![0.0, 1.0]];
        let b = vec![vec![5.0, 3.0], vec![2.0, 4.0]];
        let c = MatrixOps::matrix_multiply(&a, &b);

        assert!((c[0][0] - 5.0).abs() < 1e-10);
        assert!((c[0][1] - 3.0).abs() < 1e-10);
        assert!((c[1][0] - 2.0).abs() < 1e-10);
        assert!((c[1][1] - 4.0).abs() < 1e-10);
    }

    #[test]
    fn test_matrix_multiply_general() {
        let a = vec![vec![1.0, 2.0], vec![3.0, 4.0]];
        let b = vec![vec![5.0, 6.0], vec![7.0, 8.0]];
        let c = MatrixOps::matrix_multiply(&a, &b);

        // [1*5+2*7, 1*6+2*8] = [19, 22]
        // [3*5+4*7, 3*6+4*8] = [43, 50]
        assert!((c[0][0] - 19.0).abs() < 1e-10);
        assert!((c[0][1] - 22.0).abs() < 1e-10);
        assert!((c[1][0] - 43.0).abs() < 1e-10);
        assert!((c[1][1] - 50.0).abs() < 1e-10);
    }

    #[test]
    fn test_lu_decomposition() {
        let a = vec![
            vec![4.0, 3.0, 1.0],
            vec![3.0, 5.0, 2.0],
            vec![1.0, 2.0, 3.0],
        ];

        let (l, u) = MatrixOps::lu_decomposition(&a);

        // Verify L has 1s on diagonal
        for i in 0..3 {
            assert!((l[i][i] - 1.0).abs() < 1e-10);
        }

        // Verify L * U = A
        let lu = MatrixOps::matrix_multiply(&l, &u);
        for i in 0..3 {
            for j in 0..3 {
                assert!(
                    (lu[i][j] - a[i][j]).abs() < 1e-8,
                    "LU[{},{}] = {} != A[{},{}] = {}",
                    i,
                    j,
                    lu[i][j],
                    i,
                    j,
                    a[i][j]
                );
            }
        }
    }

    #[test]
    fn test_gaussian_elimination_2x2() {
        let a = vec![vec![2.0, 1.0], vec![1.0, 3.0]];
        let b = vec![8.0, 13.0];
        let x = MatrixOps::gaussian_elimination(&a, &b);

        // Solution: 2x+y=8, x+3y=13 → x=2.2, y=3.6
        assert!((x[0] - 2.2).abs() < 1e-8);
        assert!((x[1] - 3.6).abs() < 1e-8);
    }

    #[test]
    fn test_gaussian_elimination_3x3() {
        let a = vec![
            vec![3.0, 2.0, -1.0],
            vec![2.0, -2.0, 4.0],
            vec![-1.0, 0.5, -1.0],
        ];
        let b = vec![1.0, -2.0, 0.0];
        let x = MatrixOps::gaussian_elimination(&a, &b);

        // Verify solution: A*x = b
        for i in 0..3 {
            let mut sum = 0.0;
            for j in 0..3 {
                sum += a[i][j] * x[j];
            }
            assert!(
                (sum - b[i]).abs() < 1e-8,
                "A*x[{}] = {} != b[{}] = {}",
                i,
                sum,
                i,
                b[i]
            );
        }
    }

    #[test]
    fn test_lu_preserves_determinant() {
        let a = vec![
            vec![4.0, 7.0],
            vec![2.0, 6.0],
        ];

        let (l, u) = MatrixOps::lu_decomposition(&a);

        // det(A) = det(L) * det(U)
        // det(L) = 1 (unit lower triangular)
        // det(U) = product of diagonal
        let det_u = u[0][0] * u[1][1];
        let det_a = a[0][0] * a[1][1] - a[0][1] * a[1][0];

        assert!((det_a - det_u).abs() < 1e-10);
    }

    #[test]
    fn test_gaussian_elimination_singular_avoidance() {
        // Test with potential pivot issues
        let a = vec![
            vec![0.001, 1.0],
            vec![1.0, 1.0],
        ];
        let b = vec![1.0, 2.0];

        let x = MatrixOps::gaussian_elimination(&a, &b);
        assert_eq!(x.len(), 2);
        // Should complete without panic
    }

    #[test]
    fn test_matrix_multiply_rectangular() {
        let a = vec![vec![1.0, 2.0, 3.0], vec![4.0, 5.0, 6.0]];
        let b = vec![vec![7.0, 8.0], vec![9.0, 10.0], vec![11.0, 12.0]];
        let c = MatrixOps::matrix_multiply(&a, &b);

        // c[0][0] = 1*7 + 2*9 + 3*11 = 58
        // c[1][1] = 4*8 + 5*10 + 6*12 = 154
        assert!((c[0][0] - 58.0).abs() < 1e-10);
        assert!((c[1][1] - 154.0).abs() < 1e-10);
    }
}
